### PR TITLE
TSDF: `extractStateIntervals` method to identify an "event_time" range where state is constant

### PR DESCRIPTION
**Pull request by R7L208**
Makes progress on issue #213 

This PR defines the method `extractStateIntervals` which allows users to pass metrics which should be tracked for state changes and a definition for constant state over time across these metrics.

The returned TSDF object will contain the partition columns and a timestamp column of `StructType` with a `start` and `end` key identifying the start and end values of constant state based on the inputted metric columns and state definition.

This allows the user to perform a join with the input TSDF for the `extractStateIntervals` method to identify and isolate records of particular states.

### CODE REVIEW

File: python/tests/base.py [hunk 1]
```diff
@@ -203,6 +203,5 @@ def assertDataFramesEqual(self, dfA, dfB):
         sortedB = dfB.select(colOrder)
         # must have identical data
         # that is all rows in A must be in B, and vice-versa
-
```
Comment (tnixon) [inline python/tests/base.py:206]: Is this just a formatting change?

**Review by tnixon — COMMENTED** [review]
The changes to the example code (dlt_tempo.py and financial_services_quickstart.py) - are these just formatting changes? It doesn't appear that anything material has changed here. Are these necessary in this PR?

**Comment by R7L208** [comment]
> The changes to the example code (dlt_tempo.py and financial_services_quickstart.py) - are these just formatting changes? It doesn't appear that anything material has changed here. Are these necessary in this PR?

@tnixon - These are just formatting changes from running black at the project root. I hadn't noticed they were added but can revert these files if needed

File: python/tests/base.py (hunk 1 above)
Comment (R7L208) [reply python/tests/base.py:206]: Yes, this is just a formatting change caused by black

**Comment by tnixon** [comment]
No problem - I just wanted to make sure there wasn't other changes in these files. 
In general I prefer to keep PR changes isolated to the areas relevant for the change - otherwise it's hard to review. We can always merge formatting and unrelated things in a separate PR. 

All that said, I've certainly violated this rule myself more than once, so... 😝

**Comment by R7L208** [comment]
> No problem - I just wanted to make sure there wasn't other changes in these files. In general I prefer to keep PR changes isolated to the areas relevant for the change - otherwise it's hard to review. We can always merge formatting and unrelated things in a separate PR.
> 
> All that said, I've certainly violated this rule myself more than once, so... 😝

Reverted the changes to `examples/.`

File: python/tempo/tsdf.py [hunk 2]
```diff
@@ -1173,86 +1173,64 @@ def constantMetricState(
...
-                ">",
-                "<",
-                ">=",
-                "<=",
-            ):
+            if state_definition not in ("=", "<=>", "!=", "<>", ">", "<", ">=", "<="):
                 logger.warning(
                     "A `state_definition` which has not been tested was"
-                    "provided to the `constantMetricState` function."
+                    "provided to the `constantMetricState` method."
```
Comment (tnixon) [inline python/tempo/tsdf.py:1179]: method name has changed!

**Review by tnixon — CHANGES_REQUESTED** [review]

File: python/tempo/tsdf.py (hunk 2 above)
Comment (tnixon) [reply python/tempo/tsdf.py]: I updated it.

**Review by tnixon — APPROVED** [review]

**Pull request merged**